syntax = "proto3";

package ads;

message RealEstateAd {
    string id = 1;
    string title = 2;
    string description = 3;
    string address = 4;
    string property_id_nma = 5;
    double price = 6;
    string type = 7;
    string status = 8;
    string phone_number = 9;
    string listed_by = 10;
}

message SingleAdRequest {
    string id = 1;
    string property_id_nma = 2;
    string phone_number = 3;
}

message FilterAdsRequest {
    string property_id_nma = 1;
    string type = 2;
    string status = 3;
    double min_price = 4;
    double max_price = 5;
    int32 page = 6;
}

message MultipleAdsResponse {
    repeated RealEstateAd ads = 1;
}

message StatusResponse {
    bool success = 1;
}

message CountResponse {
    int32 count = 1;
}

service AdService {
    rpc CreateAd(RealEstateAd) returns (RealEstateAd);
    rpc GetAd(SingleAdRequest) returns (RealEstateAd);
    rpc GetAds(FilterAdsRequest) returns (MultipleAdsResponse);
    rpc CountAds(FilterAdsRequest) returns (CountResponse);
    rpc UpdateAd(RealEstateAd) returns (RealEstateAd);
    rpc DeleteAd(SingleAdRequest) returns (StatusResponse);
    rpc HasWriteAccess(SingleAdRequest) returns (StatusResponse);
}
